lib/signs/utilities/messages.ex

```diff
@@ -258,6 +304,73 @@ defmodule Signs.Utilities.Messages do
     end
   end
 
+  @spec overnight_period?(
+          DateTime.t(),
+          DateTime.t(),
+          DateTime.t(),
+          DateTime.t(),
+          boolean()
+        ) :: boolean()
+  defp overnight_period?(
+         _current_time,
+         first_scheduled_departure,
+         last_scheduled_departure,
+         _most_recent_departure,
+         false
+       )
+       when first_scheduled_departure == nil or last_scheduled_departure == nil,
+       do: false
+
+  defp overnight_period?(
+         current_time,
+         first_scheduled_departure,
+         last_scheduled_departure,
+         _most_recent_departure,
+         false
+       ),
+       do:
+         calculate_overnight_period(
+           last_scheduled_departure,
+           current_time,
+           first_scheduled_departure
+         )
+
+  defp overnight_period?(
+         current_time,
+         first_scheduled_departure,
+         last_scheduled_departure,
+         most_recent_departure,
+         true
+       ) do
+    if Timex.after?(most_recent_departure, last_scheduled_departure) do
+      most_recent_departure
+    else
+      last_scheduled_departure
+    end
+    |> calculate_overnight_period(
+      current_time,
+      first_scheduled_departure
+    )
+  end
+
+  defp calculate_overnight_period(
+         nil,
+         _current_time,
+         _first_scheduled_departure
+       ),
+       do: false
+
+  defp calculate_overnight_period(
+         last_time_to_check,
+         current_time,
+         first_scheduled_departure
+       ) do
+    # Overnight, schedules might switch to the next day
+    # so we just check if we're before the AM period, or if we're after the last time we need to check
+    Timex.after?(current_time, Timex.shift(last_time_to_check, minutes: @overnight_buffer)) ||
+      before_early_am_threshold?(current_time, first_scheduled_departure)
```

Comment:
Just checking my understanding: The API automatically flips schedules over at 3AM, so we'd transition from the first part of the `||` being `true` to the second part being `true` atomically, and so we'll remain in the overnight period across that boundary. Right?